text/0000-iri-v181-transaction-and-bundle/0000-iri-v181-transaction-and-bundle.md

```diff
@@ -0,0 +1,980 @@
++ [Bundles and transactions](https://docs.iota.org/docs/dev-essentials/0.1/concepts/bundles-and-transactions)
++ [Structure of a transaction](https://docs.iota.org/docs/dev-essentials/0.1/references/structure-of-a-transaction)
++ [Structure of a bundle](https://docs.iota.org/docs/dev-essentials/0.1/references/structure-of-a-bundle)
++ [Security levels](https://docs.iota.org/docs/dev-essentials/0.1/references/security-levels)
+
+# Motivation
+
+IOTA is a transaction settlement and data transfer layer for the Internet of Things. Messages on the network are
+`Bundle`s of one or more `Transaction`s, which in turn are sent one at a time and are stored in a distributed ledger
+called the *Tangle*. Each transaction encodes data such as sender and receiver addresses, referenced transactions in
+the Tangle, bundle hash, timestamps, and other information required to verify and process each transaction. With
+`Transaction` and `Bundle` representing the units of communication within the IOTA network, we need to be able to
+construct and represent them in memory.
+
+The current design proposal intends to be as simple as possible, provide an idiomatic Rust interface, and use existing
+IOTA terminology where appplicable. The present types are not intended to be used by end users, but rather as building
+blocks for higher level APIs. As such, the different kinds of transactions are not encoded as Rust types, and the bundle
+builders do not contain any logic to implicitly push more transactions into their stack. For a bundle to be buildable,
+all required transactions have to be present when validating and building. Otherwise the build will fail.
+
+As an example, a simple `Bundle` that intends to transfer funds, withdrawing tokens from one address and depositing them
+in another would be constructed like this:
+
+```rust
+let mut outgoing_bundle_builder = OutgoingBundleBuilder::new();
+
+let mut withdrawing_tx_builder = TransactionBuilder::new();
+let mut depositing_tx_builder = TransactionBuilder::new();
+
+// The transactions here only need to set addresses, values, and tags. All other fields, such as the signature
+// of the withdrawing transaction, will be filled in by the bundle builder.
+withdrawing_tx_builder
+    .address(address_providing_funds)
+    .value(value_to_transfer)
+    .tag(custom_tag);
+
+depositing_tx_builder
+    .address(address_receiving_funds)
+    .value(value_to_transfer)
+    .tag(custom_tag);
+
+let outgoing_bundle = outgoing_bundle_builder
+    .push(withdrawing_tx_builder)
+    .push(depositing_tx_builder)
+    // Seal the bundle, adding indices to its transactions and assigning a bundle hash. This also prevents
+    // pushing more transactions into the bundle.
+    .seal()?
+    // Withdrawal transactions are signed, providing some signature scheme, seed, and wallet. If none are present, this
+    // step can be skipped.
+    .sign(signature_scheme, seed, wallet)?
```

Comment:
Does this suggest that multiple signature schemes are allowed like WOTS/ECC/...
If so, is it really needed?

Reply:
That depends on how we implement `sign`. If `sign` takes a `S: SignatureScheme`, then yeah --- it would allow for different schemes. If it takes a `WOTS`, then no. Either way, we'd probably pass the scheme in explicitly because it might need to be configured.

But let's see what its RFC brings.